### PR TITLE
Added http.Filesystem as a source

### PR DESCRIPTION
**Pull request by mikijov**
This is a simpler version of #144. I have added http.Filesystem as a source, which has to be initialized with an instance of a http.Filesystem. Caller has to use WithInstance() passing http.Filesystem to use. This source cannot be used with Open() as this would unnecessarily require adding many dependencies. I am providing this as an alternative to #144 until that PR is properly prepared.

### CODE REVIEW

File: source/http_filesystem/http_filesystem.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	"net/http"
+	"os"
+	"path"
+	"sort"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+func init() {
+	source.Register("http-fs", &httpFs{})
```
Comment (dhui) [inline source/http_filesystem/http_filesystem.go]: Don't register the source since you can't use `source.Open()` with this "driver" anyways.

File: source/http_filesystem/http_filesystem.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+type Config struct {
+	Dir string
+}
+
+type httpFs struct {
```
Comment (dhui) [inline source/http_filesystem/http_filesystem.go]: Export this as `HTTPFS`

File: source/http_filesystem/http_filesystem.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+		if !s.migrations.Append(m) {
+			return nil, fmt.Errorf("unable to parse file %v", fi.Name())
+		}
+	}
+
+	return s, nil
+}
+
+func (f *httpFs) Open(url string) (source.Driver, error) {
```
Comment (dhui) [inline source/http_filesystem/http_filesystem.go]: Remove the `Open()` receiver method

File: source/http_filesystem/http_filesystem_test.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
+package httpfs
```
Comment (dhui) [inline source/http_filesystem/http_filesystem_test.go]: Use a test struct that embeds `HTTPFS` and has an `Open()` receiver method

File: source/http_filesystem/http_filesystem.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
+package httpfs
```
Comment (dhui) [inline source/http_filesystem/http_filesystem.go]: rename the folder and file to be the same as the package name. e.g. `http_filesystem` -> `httpfs`

File: source/http_filesystem/http_filesystem_test.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+
+	st "github.com/golang-migrate/migrate/v4/source/testing"
+)
+
+func Test(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "")
```
Comment (dhui) [inline source/http_filesystem/http_filesystem_test.go]: Use [bytes.Buffer](https://golang.org/pkg/bytes/#Buffer) instead of a temp dir since httpfs doesn't necessarily work with local files.

File: source/http_filesystem/http_filesystem_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+	for i := 0; i < 1000; i++ {
+		mustWriteFile(t, tmpDir, fmt.Sprintf("%v_foobar.up.sql", i), "")
+		mustWriteFile(t, tmpDir, fmt.Sprintf("%v_foobar.down.sql", i), "")
+	}
+
+	return tmpDir
+}
+
+func BenchmarkOpen(b *testing.B) {
```
Comment (dhui) [inline source/http_filesystem/http_filesystem_test.go]: Benchmarks don't seem necessary

File: source/http_filesystem/http_filesystem.go (hunk 1 above)
Comment (mikijov) [reply source/http_filesystem/http_filesystem.go]: Done.

**Comment by mikijov** [comment]
More changes coming...

**Comment by nkonev** [comment]
@dhui seems all reviews outdated

File: source/http_filesystem/http_filesystem.go (hunk 3 above)
Comment (dhui) [reply source/http_filesystem/http_filesystem.go]: This still needs to be done

File: source/http_filesystem/http_filesystem.go (hunk 5 above)
Comment (dhui) [reply source/http_filesystem/http_filesystem.go]: Still need to rename the file

File: go.mod [hunk 8]
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/go-sql-driver/mysql v1.4.1
 	github.com/go-stack/stack v1.8.0 // indirect
 	github.com/gocql/gocql v0.0.0-20181124151448-70385f88b28b
+	github.com/golang-migrate/migrate v3.5.4+incompatible
```
Comment (dhui) [inline go.mod]: Why is this here?

File: source/httpfs/http_filesystem.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	return v, nil
+}
+
+// ReadUp is implementation of source.ReadUp(). Please see it for documentation.
+func (f *HTTPFS) ReadUp(version uint) (r io.ReadCloser, identifier string, err error) {
+	if m, ok := f.migrations.Up(version); ok {
+		p := path.Join(f.path, m.Raw)
+		r, err := f.fs.Open(p)
+		if err != nil {
+			return nil, "", fmt.Errorf("can't open file %s: %s", p, err)
```
Comment (dhui) [inline source/httpfs/http_filesystem.go]: Don't mask the original error and instead use `go-multierror`

File: source/httpfs/http_filesystem.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+	return nil, "", &os.PathError{Op: fmt.Sprintf("read version %v", version), Path: f.path, Err: os.ErrNotExist}
+}
+
+// ReadDown is implementation of source.ReadDown(). Please see it for documentation.
+func (f *HTTPFS) ReadDown(version uint) (r io.ReadCloser, identifier string, err error) {
+	if m, ok := f.migrations.Down(version); ok {
+		p := path.Join(f.path, m.Raw)
+		r, err := f.fs.Open(p)
+		if err != nil {
+			return nil, "", fmt.Errorf("can't open file %s: %s", p, err)
```
Comment (dhui) [inline source/httpfs/http_filesystem.go]: ditto

File: source/httpfs/http_filesystem.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// Open is not implemented as many FileSystems cannot be opened or have
+// custom initialization.
+//
+// Common use:
+//   plainFileSystem, err := WithInstance(http.Dir(tmpDir), &Config{})
+//
+//   escEmbeddedFileSystem, err := WithInstance(FS(false), &Config{})
+func (f *HTTPFS) Open(url string) (source.Driver, error) {
```
Comment (dhui) [inline source/httpfs/http_filesystem.go]: Needs to be removed as mentioned [here](https://github.com/golang-migrate/migrate/pull/163#discussion_r252909158)

File: source/httpfs/http_filesystem_test.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+	for i := 0; i < 1000; i++ {
+		mustWriteFile(t, tmpDir, fmt.Sprintf("%v_foobar.up.sql", i), "")
+		mustWriteFile(t, tmpDir, fmt.Sprintf("%v_foobar.down.sql", i), "")
+	}
+
+	return tmpDir
+}
+
+func TestEscFilesystem(t *testing.T) {
```
Comment (dhui) [inline source/httpfs/http_filesystem_test.go]: Remove this test

File: source/httpfs/http_filesystem_test.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	"path"
+	"path/filepath"
+	"testing"
+
+	"github.com/golang-migrate/migrate/v4/source/httpfs/escfs"
+	st "github.com/golang-migrate/migrate/v4/source/testing"
+)
+
+func Test(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "")
```
Comment (dhui) [inline source/httpfs/http_filesystem_test.go]: As mentioned [here](https://github.com/golang-migrate/migrate/pull/163#discussion_r252911552), don't use a a temp dir for testing. You can use `afero` instead.

File: source/httpfs/escfs/escfs.go [hunk 14]
```diff
@@ -0,0 +1,332 @@
+// Code generated by "esc -o escfs/escfs.go -pkg escfs -prefix escfs/ escfs/"; DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (mikijov) [reply source/httpfs/escfs/escfs.go]: This is not support for ESC. Once generated ESC is not being used any more. This directory is just test data. This is how embedded filesystems work.

File: source/httpfs/http_filesystem_test.go [hunk 15]
```diff
@@ -1,3 +1,4 @@
+//go:generate esc -o escfs/escfs.go -pkg escfs -prefix escfs/ escfs/
```
[the remark this answers is not in the document]
Comment (mikijov) [reply source/httpfs/http_filesystem_test.go]: It is referenced only for code generation. It is not used for build or test. This is just a helper for anyone wanting to change tests that use embedded file system. Anyone who does not change ESC test files will not be using this nor would it affect their build.

**Comment by mikijov** [comment]
I will have to come back to this when I have bit more time...

File: source/httpfs/http_filesystem_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (dhui) [reply source/httpfs/http_filesystem_test.go]: I'd like to avoid code-gen if possible. It seems like the same testing needs can be met by `afero`. https://github.com/spf13/afero#using-afero-for-testing

**Comment by dio** [comment]
@mikijov do you think you need help on this? Thanks!

**Comment by dhui** [comment]
Closed in favor of https://github.com/golang-migrate/migrate/pull/293
